### PR TITLE
MPP-3698: Enable `mypy --check-untyped-def --no-implicit-reeexport` across code

### PR DESCRIPTION
**Pull request by jwhitlock**
For MPP-3698, I updated the code so both `--check-untyped-def` is used across our code. This means that code without type annotations is checked by `mypy` rather than skipped, and finds errors like not considering `None` or other alternate return types.

This is a big PR, but the changes themselves are small.

## Bugs Fixed

* `api/authentication.py`
  -  `introspect_token` - Error handler referred to `fxa_resp`, was not initialized (Sentry `RELAY-J0`, `RELAY-J2`)
* `api/permissions.py`
  - `HasPremium` called a `bool` (unused in code)
* `emails/management/commands/process_emails_from_sqs.py`
  - `Command.handle` - Use different name for exception inside of an exception handler
  
## Code changes
* Moved app config helper functions to app file. For example `emails_config()` is now in `emails.app.py`.
* Lots of new annotations
* Add `assert` to ensure a value is not `None` when `mypy` complained
* Add `isinstance` to ensure `request.user` is not an anonymous user in API code
* Add `hasattr` checks as needed for dynamically added fields
* Comments and long code lines wrapped to 88 chars (in changed files). Long function names have a `# noqa: E501` comment.
* Used walrus operator (`:=`) to assign and assert (usually not `None`) in one line
* Changes `Optional[X]` to Python 3.9 format `X | None`
* More use of `from __future__ import annotations` for [PEP-563](https://peps.python.org/pep-0563/), Postponed Evaluation of Annotations, to allow more forward annotations without quotes.
* `api/tests/authentication_tests.py`
  - Used instance of `FxaTokenAuthentication` instead of class in `FxaTokenAuthenticationTest`
  - Used Django 4.2 header-setting style in `FxaTokenAuthenticationTest`
* `privaterelay/settings.py` 
  - `RELAY_CHANNEL` is now restricted to "local", "dev", "stage", or "prod"
  - Added many annotations to `privaterelay/settings.py` for "circular import" issues. 
  - Removed `cast=str` when it is the default.
  - Set `AUTO_RELOAD_BUNDLES = False` to avoid `django-ftl` warning

### CODE REVIEW

**Comment by jwhitlock** [comment]
PR #4321 is merged, this is ready for review.

File: privaterelay/settings.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 
 from __future__ import annotations
 from pathlib import Path
-from typing import Any, Optional, TYPE_CHECKING
+from typing import Any, Literal, Optional, TYPE_CHECKING, cast, get_args
```
Comment (groovecoder) [inline privaterelay/settings.py:15]: suggestion (non-blocking): change L719 to `sentry_release: str | None` and remove import of `Optional`

File: privaterelay/settings.py [hunk 2]
```diff
@@ -226,7 +233,7 @@
 
 DJANGO_STATSD_ENABLED = config("DJANGO_STATSD_ENABLED", False, cast=bool)
 STATSD_DEBUG = config("STATSD_DEBUG", False, cast=bool)
-STATSD_ENABLED = DJANGO_STATSD_ENABLED or STATSD_DEBUG
+STATSD_ENABLED: bool = DJANGO_STATSD_ENABLED or STATSD_DEBUG
 STATSD_HOST = config("DJANGO_STATSD_HOST", "127.0.0.1")
 STATSD_PORT = config("DJANGO_STATSD_PORT", "8125")
 STATSD_PREFIX = config("DJANGO_STATSD_PREFIX", "fx.private.relay")
```
Comment (groovecoder) [inline privaterelay/settings.py:239]: question (non-blocking): I can't figure out why we declare the type of `STATSD_ENABLED` but none of these other ones? Does the default value in the `config` call effectively declare the type for `mypy`? I can't tell what the pattern is of when we do/don't declare the type of a `settings` variable.

File: privaterelay/tests/fxa_utils_tests.py [hunk 3]
```diff
@@ -65,16 +63,13 @@ def test_get_phone_subscription_dates_subscription_not_in_data_no_free_phone(
     )
 
 
+@override_flag("free_phones", active=True)
```
Comment (groovecoder) [inline privaterelay/tests/fxa_utils_tests.py:66]: praise: nice clean-up.

File: privaterelay/tests/mgmt_sync_phone_related_dates_on_profile_tests.py [hunk 4]
```diff
@@ -85,13 +86,12 @@ def test_free_phone_user_gets_only_date_phone_subscription_reset_field_updated(
...
     expected_now = patch_datetime_now
     mocked_dates.return_value = (None, None, None)
-    account = baker.make(SocialAccount, provider="fxa")
+    account: SocialAccount = baker.make(SocialAccount, provider="fxa")
     profile = Profile.objects.get(user=account.user)
     baker.make(Flag, name="free_phones")
-    free_phones_flag = Flag.objects.filter(name="free_phones").first()
+    free_phones_flag = Flag.objects.get(name="free_phones")
     free_phones_flag.users.add(profile.user)
     free_phones_flag.save()
```
Comment (groovecoder) [inline privaterelay/tests/mgmt_sync_phone_related_dates_on_profile_tests.py:94]: suggestion (non-blocking): this block is used at least twice. Can we move this logic into a `get_user_free_phones()` helper function?

File: pyproject.toml [hunk 5]
```diff
@@ -55,40 +55,3 @@ module = [
...
-    "emails.tests.models_tests",
-    "emails.tests.views_tests",
-    "emails.utils",
-    "phones.models",
-    "phones.tests.models_tests",
-    "privaterelay.tests.fxa_utils_tests",
-    "privaterelay.tests.mgmt_sync_phone_related_dates_on_profile_tests",
-    "privaterelay.tests.mgmt_update_phone_remaining_stats_tests",
-    "privaterelay.tests.views_tests",
-]
```
Comment (groovecoder) [inline pyproject.toml:94]: praise: woohoo!! look at all these modules with better type-checking now!

**Review by groovecoder — CHANGES_REQUESTED** [review]
One blocking question about returning `401` status code that's very important to get right because Firefox clients change their behavior based on that status code.

**Review by jwhitlock — COMMENTED** [review]
Thanks for the quick review @groovecoder! I've answered some questions, and made two suggested changes:

* Removed `Optional` from `privaterelay/settings.py`
* Added `create_free_phones_flag_for_user` helper

File: api/authentication.py [hunk 6]
```diff
@@ -85,18 +86,18 @@ def get_fxa_uid_from_oauth_token(token, use_cache=True):
         raise APIException("Did not receive a 200 response from FXA.")
 
     if not fxa_resp_data["json"].get("active"):
-        raise AuthenticationFailed("FXA returned active: False for token.", code=401)
+        raise AuthenticationFailed("FXA returned active: False for token.")
```
[the remark this answers is not in the document]
Comment (groovecoder) [reply api/authentication.py:89]: Thanks for the detailed info! That, and the fact that `test_inactive_token_responds_with_401` test makes sure the `401` is returned in the right case, and that test is still passing, makes me non-blocking on this again!

**Review by groovecoder — APPROVED** [review]

File: api/authentication.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (say-yawn) [reply api/authentication.py:89]: I found out that `code` is returned in response data like:
```python
> response.data
{'detail': ErrorDetail(string='FXA returned active: False for token.', code='authentication_failed')}
```
And I checked in the [Firefox code](https://searchfox.org/mozilla-central/rev/3c72de9280ec57dc55c24886c6334d9e340500e8/toolkit/components/satchel/integrations/FirefoxRelay.sys.mjs#108) @groovecoder mentioned and it does not refer to `code` in the data details. So I also think we are good.

**Pull request merged**